### PR TITLE
feat: interfaces for async embeddings, implement async openai

### PR DESCRIPTION
**Pull request by tyree731**
This change adds support to the base `Embeddings` class for two methods, `aembed_query` and `aembed_documents`, those two methods supporting async equivalents of `embed_query` and
`embed_documents` respectively. This ever so slightly rounds out async support within langchain, with an initial implementation of this functionality being implemented for openai.

Implements #6109

### CODE REVIEW

**Comment by tyree731** [comment]
CC @hwchase17 @dev2049 for review. the `_aget_len_safe_embeddings` bit is a bit rough, as I couldn't think of a great way to reuse the non-embedding work from `_get_len_safe_embeddings`, open to suggestions there.

**Comment by tyree731** [comment]
Let me know if there is anything I can do to help push this along. I know you're all busy, but just want to make sure there isn't anything on my end holding this up. Also:

- I wasn’t 100% on where I should be adding more documentation, so if I should be adding more documentation, kindly point me in the right direction there.
- I added integration tests for openai async embeddings, but if I should be adding tests for fake embeddings async or some such, I can do so.

Thanks!

File: langchain/embeddings/openai.py [hunk 1]
```diff
@@ -53,6 +54,38 @@ def _create_retry_decorator(embeddings: OpenAIEmbeddings) -> Callable[[Any], Any
...
+            | retry_if_exception_type(openai.error.ServiceUnavailableError)
+        ),
+        before_sleep=before_sleep_log(logger, logging.WARNING),
+    )
+
+    def wrap(func: Callable) -> Callable:
+        async def wrapped_f(*args: Any, **kwargs: Any) -> Callable:
+            async for _ in async_retrying:
+                return await func(*args, **kwargs)
+            assert False, "this is unreachable"
```
Comment (eyurtsev) [inline langchain/embeddings/openai.py:82]: nit: more common to reserve assert statements for the test code and instead raise AssertionErrors

**Comment by eyurtsev** [comment]
Hi @tyree731 , code looks good -- let's remove the abstractmethod to reduce scope and avoid breaking changes and then code is good to merge.

File: langchain/embeddings/base.py [hunk 2]
```diff
@@ -10,6 +10,14 @@ class Embeddings(ABC):
     def embed_documents(self, texts: List[str]) -> List[List[float]]:
         """Embed search docs."""
 
+    @abstractmethod
```
[the remark this answers is not in the document]
Comment (tyree731) [reply langchain/embeddings/base.py]: Makes sense, updated to raise NotImplementedError in the base class.

File: langchain/embeddings/openai.py (hunk 1 above)
Comment (tyree731) [reply langchain/embeddings/openai.py]: Fixed, thanks

**Comment by tyree731** [comment]
Updated based on your comments. Let me know if there's anything I need to do for the vercel failure.

**Review by hwchase17 — COMMENTED** [review]
we can add these methods to the base class, but lets just do something like

```
    async def aembed_documents(
        self, texts: List[str], chunk_size: Optional[int] = 0
    ) -> List[List[float]]:
    raise NotImplementedError
```
In the base class (rather than mark as abstract and have each subclass have to implement that)

**Comment by tyree731** [comment]
Sorry for the delay in responding @hwchase17 , currently on vacation in Germany for the next couple of weeks, so I'll get to your comments when I get back.

**Pull request closed without merging**